pkg/operatorsource/updater.go

```diff
@@ -0,0 +1,113 @@
+	"github.com/operator-framework/operator-marketplace/pkg/phase"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// Updater is an interface that can be used to check whether a remote registry
+// has any update(s) and trigger a rebuild of in cluster OperatorSource cache.
+type Updater interface {
+	// Check contacts the remote registry associated with the specified
+	// OperatorSource object, fetches release metadata and determines whether
+	// the remote registry has new update(s).
+	//
+	// It returns true if the remote registry has update(s), otherwise it
+	// returns false.
+	// If the function encounters any error then it returns (false, err).
+	Check(source *datastore.OperatorSourceKey) (bool, error)
+
+	// Trigger triggers a rebuild of the cache associated with the
+	// specified OperatorSource object.
+	//
+	// It fetches the latest copy of the specified OperatorSource object and
+	// then sets the phase to 'Purging' so that the cache is invalidated and
+	// reconciliation can start new.
+	//
+	// On return, deleted is set to true if the object has already been deleted.
+	// updateErr is set to the error the function encounters while it tries
+	// to update the OperatorSource object.
+	Trigger(source *datastore.OperatorSourceKey) (deleted bool, updateErr error)
+}
+
+// updater implements the Updater interface.
+type updater struct {
+	factory      appregistry.ClientFactory
+	datastore    datastore.Writer
+	client       client.Client
+	transitioner phase.Transitioner
+}
+
+func (u *updater) Check(source *datastore.OperatorSourceKey) (bool, error) {
+	// Get the latest version of the operator source from underlying datastore.
+	source, exists := u.datastore.GetOperatorSource(source.UID)
+	if !exists {
+		return false, errors.New("The given OperatorSource object does not exist in datastore")
+	}
+
+	registry, err := u.factory.New(source.Spec.Type, source.Spec.Endpoint)
+	if err != nil {
+		return false, err
+	}
+
+	metadata, err := registry.ListPackages(source.Spec.RegistryNamespace)
+	if err != nil {
+		return false, err
+	}
+
+	updated, err := u.datastore.OperatorSourceHasUpdate(source.UID, metadata)
+	if err != nil {
+		return false, err
+	}
+
+	return updated, nil
+}
+
+func (u *updater) Trigger(source *datastore.OperatorSourceKey) (deleted bool, updateErr error) {
+	instance := &v1alpha1.OperatorSource{}
+
+	// Get the current state of the given object before we make any decision.
+	if err := u.client.Get(context.TODO(), source.Name, instance); err != nil {
+		// Not found, the given OperatorSource object could have been deleted.
+		// Treat it as no error and indicate that the object has been deleted.
+		if k8s_errors.IsNotFound(err) {
+			deleted = true
+			return
+		}
+
+		// Otherwise, it is an error.
+		updateErr = err
+		return
+	}
+
+	// Needed because sdk does not get the gvk.
+	instance.EnsureGVK()
+
+	if instance.GetCurrentPhaseName() == phase.OperatorSourceDownloading {
```

Comment:
Is checking for `OperatorSourceDownloading` sufficient? Shouldn't any sort of sync happen only if the status is `Succeeded` or `Failed`?

Reply:
Ideally we don't want to rely on phase transition logic. It introduces a coupling I think. If the object is stuck at a particular phase then we would not pull update(s).  
But the downside is on rare occasion we might pull update(s) immediately after a fresh download.
In reality, if there are too many updates of an object within a short amount time, k8s will discard the older version(s) and will inform of the latest version only.

Reply:
>Ideally we don't want to rely on phase transition logic. It introduces a coupling I think. 

That is what I was hoping for but I thought you wanted that coupling.

>If the object is stuck at a particular phase then we would not pull update(s).

Why should we pull updates if it is stuck in a state? Are you thinking of using this thread as a kicker to make it keep moving as a side effect? I am not a big fan of that side effect.

>In reality, if there are too many updates of an object within a short amount time, k8s will discard the older version(s) and will inform of the latest version only.

I don't see how this relates to my original question.

Reply:
* I am happy to remove the check for `Downloading` phase. I added this based on your question `what if a download is in progress while we are kicking off an datastore update`.

* I believe this is not a side effect, if you have new update(s) it's time to reconcile the OperatorSource CR irrespective of what phase it is in. 

* I mention this because k8s internally optimizes and informs of the latest version only in between two reconciliation window.

Reply:
I am not asking for the check to be removed. My point is that if the `OperatorSource` is in any phase other than `Succeeded` or `Failed` we should not pull down updates. Instead we back off and grab the updates during the next sync period. So the check should reflect that.

>I believe this is not a side effect, if you have new update(s) it's time to reconcile the OperatorSource CR irrespective of what phase it is in.

If a particular phase (not `Succeeded` or `Failed`) is having an issue preventing it from moving to the next, there is a potential for it to keep going back and forth between the erroneous phase and `Purging` every sync cycle. It is better at that point for the admin to intervene.

Reply:
> there is a potential for it to keep going back and forth between the erroneous phase and `Purging` every sync cycle. It is better at that point for the admin to intervene.

* A sync cycle is not designed to cause `back and forth`. Can you give me a scenario?
* Even if there is potential for `back and forth`, this function is not the place to address it. This function does one thing - if there are update(s) for an `OperatorSource`, request for reconciliation to start anew. That's what I meant when I referred to less coupling, this function should not care what phase the object is in. It's intention is to request a reconciliation.
* If there are update(s), sync cycle should be able to trigger reconciliation, without need of any human intervention. It's safe not to assume that the `OperatorSource` object will be in certain phase(s) in an asynchronous world.

I summarized my points, if you still want the change, I will add the check.

Reply:
>A sync cycle is not designed to cause back and forth. Can you give me a scenario?

I gave you a generic scenario in my previous response. There could be an error during validation causing it go to failed. The sync cycle will set it back to purging and it will again fail in validation. This will keep happening on very sync cycle unless the root cause of the validation failure is not fixed by updating the CR spec. Yes, there is no harm in this happening but do we need this churn?

>That's what I meant when I referred to less coupling, this function should not care what phase the object is in. It's intention is to request a reconciliation.

If that is your desired effect then there should be no check.

>If there are update(s), sync cycle should be able to trigger reconciliation, without need of any human intervention. It's safe not to assume that the OperatorSource object will be in certain phase(s) in an asynchronous world.

I am not saying a sync cycle should require human intervention nor am I saying that we should assume the object should be in a particular phase. I am just calling out that the sync cycle can cause an object to keep going back and forth between two phases without it ever succeeding. I hope my above example helped clarify this. 

Any how I think it is best we table this discussion for now and address it separately if needed. I will approve this as it stands or with an updated check or without a check. I leave that to you.